plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiOutputExtension.kt

```diff
@@ -171,19 +154,22 @@ class JvmAbiOutputExtension(
                                 return abiMetadataProcessor(delegate)
                             }
 
-                            override fun visitEnd() {}
-                        }, 0)
+                            override fun visitEnd() {
+                                keptFields.sortedWith(compareBy(FieldNode::name, FieldNode::desc)).forEach { it.accept(cv) }
+                                keptMethods.sortedWith(compareBy(MethodNode::name, MethodNode::desc)).forEach { it.accept(cv) }
 
-                        innerClassesToKeep.addInnerClasses(innerClassInfos, internalName)
-                        innerClassesToKeep.addOuterClasses(innerClassInfos)
+                                innerClassesToKeep.addInnerClasses(innerClassInfos, internalName)
+                                innerClassesToKeep.addOuterClasses(innerClassInfos)
```

Comment:
What's going on here? 

We're adding same `innerClassInfos` collection to both `addInnerClasses` and `addOuterClasses`, this looks quite sketchy and unmaintainable…

Reply:
Sounds like you got confused by the JVMS attribute name! "InnerClasses" doesn't mean that it lists the inner classes of the current class, it means that it lists relationships of the "class X is inner to class Y" kind. On top of classes that are inner to the current class, it should also contain such entries for all *outer* classes of the current class, classes referenced in descriptors, and class constants in the constant pool.

```
class A {
  class B
}

class C {
  class D {
    class E {
      class F

      fun foo() = A.B::class.java

      // InnerClasses of E:
      //   D=C$D of C         [marked by addOuterClasses]
      //   E=C$D$E of C$D     [marked by addOuterClasses]
      //   F=C$D$E$F of C$D$E [marked by addInnerClasses]
      //   B=A$B of A         [marked by ClassRemapper]
    }
  }
}
```